fix: Removed network call to backend server when transitioning from "Member Profile" to "Send Request" screen.

yashhalgaonkar opened the pull request:
Removed network call to backend server when transitioning from "Member Profile" to "Send Request" screen.

### Description
There was a redundant network call being made while transitioning from Member Profile Page to Send Request Page to get the current user details. The motivation was to remove that call to save device resources and provide a more seamless service to the user.

I changed the **members_page_state.dart** to include currentUser of type _User_ in _MembersPageSuccess_ state. This will have all the details of the current user.

I changed **members_page_bloc.dart** to make all the network calls to get current user details and pass that to _MembersPageSuccess_ object instance whenever it is created.

I changed **members_page.dart** to pass an instance of _MembersPageBloc_ to _MemberProfileScreen_.

I changed the **member_profile.dart** to accept the instance of _MemberPageBloc_ passed to it and use to to get current user details.


Fixes #121 

### Flutter Channel:
- [x] I have used the Flutter Beta channel on my local machine 

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)



### How Has This Been Tested?
I have tested the code manually on my emulator and device. To reproduce, just go from Member Profile to Send Requesnt Page.


### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged



**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] Any dependent changes have been published in downstream modules

yashhalgaonkar committed:
type : Enhancement

Removed network call to backend server when transitioning from "Member Profile" to "Send Request" screen.

robotjellyzone reviewed (commented):
@yashhalgaonkar can you please fill those Description & HOW THIS HAS BEEN Tested sections and also, do check the part of code/quality assurance section as well.

yashhalgaonkar commented:
@robotjellyzone I have filled the Description and Test section.

Techno-Disaster commented:
@yashhalgaonkar Just went through the PR, so basically instead of making the user network call when we click on the send request button like we did earlier, you are making the request when the page loads right?

yashhalgaonkar commented:
@Techno-Disaster Yeah. I have made network call inside the _members_page_bloc.dart_  whenever an instance of _MembersPageSuccess_ is created.

Techno-Disaster commented:
> @Techno-Disaster Yeah. I have made network call inside the _members_page_bloc.dart_ whenever an instance of _MembersPageSuccess_ is created.

The objective was reducing the extra api call, also the way you have implemented this the currentUser is fetched every time a new page is loaded in pagination, this would affect the UX, can you try and have some workaround this?

yashhalgaonkar commented:
> > @Techno-Disaster Yeah. I have made network call inside the _members_page_bloc.dart_ whenever an instance of _MembersPageSuccess_ is created.
> 
> The objective was reducing the extra api call, also the way you have implemented this the currentUser is fetched every time a new page is loaded in pagination, this would affect the UX, can you try and have some workaround this?

Should I use persistent storage to store the information of current user in local memory? That is the only workaround I can think of.

yashhalgaonkar commented:
> > @Techno-Disaster Yeah. I have made network call inside the _members_page_bloc.dart_ whenever an instance of _MembersPageSuccess_ is created.
> 
> The objective was reducing the extra api call, also the way you have implemented this the currentUser is fetched every time a new page is loaded in pagination, this would affect the UX, can you try and have some workaround this?

No. Network calls are not made every time a new page is loaded during pagination. I checked it.

techno-disaster commented on lib/screens/home/pages/members/bloc/members_page_bloc.dart:
```diff
@@ -25,47 +26,66 @@ class MembersPageBloc extends Bloc<MembersPageEvent, MembersPageState> {
-              await userRepository.getVerifiedUsers((currentState.users.length ~/ 10) + 1);
+          final users = await userRepository
+              .getVerifiedUsers((currentState.users.length ~/ 10) + 1);
+          final User currentUser = await userRepository.getCurrentUser();
```
@yashhalgaonkar I'm talking about this line?

techno-disaster replied:
this is the event added when the user scrolls to the end of the page to load more users

yashhalgaonkar replied:
Okay. I am looking into this.

yashhalgaonkar replied:
Should I use sqflite to store the current user info in local storage? This will make it easy to get current user info in all other parts of app as well, solving the problem once and for all.

techno-disaster replied:
sure, give it a try

yashhalgaonkar committed:
type : Enhancement

Removed the redundant network calls being made to fetch user when the refreshed and when scrolled to the end of the list.

yashhalgaonkar commented on lib/screens/home/pages/members/bloc/members_page_bloc.dart:
```diff
@@ -25,47 +26,67 @@ class MembersPageBloc extends Bloc<MembersPageEvent, MembersPageState> {
-          yield MembersPageSuccess(users: users, hasReachedMax: false);
+          final List<User> users =
+              await userRepository.getVerifiedUsers(pageNumber);
+          final User currentUser = await userRepository.getCurrentUser();
```
This network call will be made when the state is _MembersPageInitial_. This is a necessary call.

yashhalgaonkar commented on lib/screens/home/pages/members/bloc/members_page_bloc.dart:
```diff
@@ -25,47 +26,67 @@ class MembersPageBloc extends Bloc<MembersPageEvent, MembersPageState> {
               : MembersPageSuccess(
                   users: currentState.users + users,
                   hasReachedMax: false,
+                  currentUser: currentState.currentUser,
```
Here while initializing _MembersPageSuccess_, we need not make the network call again. So removed it.

yashhalgaonkar committed:
type : Enhancement

Removed the redundant network calls being made to fetch user when the refreshed and when scrolled to the end of the list.

yashhalgaonkar commented:
Now the network call for fetching the currentUser is made only once when state changes from _MembersPageInitial_ state to _MembersPageSuccess_ state, as it should be.

techno-disaster reviewed (approved):
LGTM

Techno-Disaster commented:
@anitab-org/qa-team can someone give this a test? you can find the apk in the github actions artifacts

rpattath commented:
@Techno-Disaster the PR needs 1 more code review approval before the QA team can test.

Techno-Disaster commented:
@anitab-org/coding-team can someone give this a quick review?
